File: 160-intersection-of-two-linked-lists/160-intersection-of-two-linked-lists.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution 
{
    public ListNode GetIntersectionNode(ListNode headA, ListNode headB) 
    {
        Dictionary<ListNode, int> nodesA = new Dictionary<ListNode, int>();
        while(headA != null)
        {
            nodesA.Add(headA, 0);
            headA = headA.next;
        }
        
        while(headB != null)
        {
            if(nodesA.ContainsKey(headB))
                return headB;
            headB = headB.next;
        }
        
        return null;
    }
}

File: 160-intersection-of-two-linked-lists/160-intersection-of-two-linked-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
#include <unordered_map>

class Solution 
{
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB)     
    {
        unordered_set<ListNode*> set;
        while(headA && headB)
        {
            if(!set.insert(headA).second)
                return headA;
            if(!set.insert(headB).second)
                return headB;
            headA = headA->next; headB = headB->next;
        }
        
        while(headA)
        {
            if(!set.insert(headA).second)
                return headA;
            headA = headA->next;
        }
        while(headB)
        {
            if(!set.insert(headB).second)
                return headB;
            headB = headB->next;
        }
        
        return nullptr;
    }
};